napari/components/_tests/test_layer_slicer.py

```diff
@@ -0,0 +1,244 @@
+#
+# behaviors we want to test
+# scheduling logic of the slicer (and not correctness of the slice response value)
+#
+# - for layers that support async, the slice task should not be run on the main thread (we don't want to block the calling)
+# - for layers that do not support async, slicing should always be done once the method returns
+# - slice requests should be run on the main thread
+# - pending tasks are cancelled (at least when the new task will slice all layers for the pending task)
+#
+# run all tests with:
+# pytest napari/components/_tests/test_layer_slicer.py -svv
+
+
+@dataclass(frozen=True)
+class FakeSliceRequest:
+    id: int = -1
+
+
+@dataclass(frozen=True)
+class FakeSliceResponse:
+    id: int = -1
+
+
+class FakeAsyncLayer:
+    def __init__(self):
+        self.slice_count = 0
+        self.lock = RLock()
+
+    def _make_slice_request(self, dims) -> FakeSliceRequest:
+        assert current_thread() == main_thread()
+        self.slice_count += 1
+        request = FakeSliceRequest(id=self.slice_count)
+        return request
+
+    def _get_slice(self, request: FakeSliceRequest) -> FakeSliceResponse:
+        assert current_thread() != main_thread()
+        with self.lock:
+            return FakeSliceResponse(id=request.id)
+
+    def _is_async(self) -> bool:
+        return True
+
+
+class FakeSyncLayer:
+    def __init__(self):
+        self.slice_count: int = 0
+
+    def _slice_dims(self, *args, **kwargs) -> None:
+        self.slice_count += 1
+
+    def _is_async(self) -> bool:
+        return False
+
+
+@pytest.fixture()
+def layer_slicer():
+    layer_slicer = _LayerSlicer()
+    yield layer_slicer
+    layer_slicer.shutdown()
+
+
+def test_slice_layers_async_with_one_async_layer(layer_slicer):
+    layer = FakeAsyncLayer()
+
+    future = layer_slicer.slice_layers_async(layers=[layer], dims=Dims())
+
+    assert future.result()[layer].id == 1
+
+
+def test_slice_layers_async_with_multiple_async_layer(layer_slicer):
+    layer1 = FakeAsyncLayer()
+    layer2 = FakeAsyncLayer()
+
+    future = layer_slicer.slice_layers_async(
+        layers=[layer1, layer2], dims=Dims()
+    )
+
+    assert future.result()[layer1].id == 1
+    assert future.result()[layer2].id == 1
+
+
+def test_slice_layers_async_emits_ready_event_when_done(layer_slicer):
+    layer = FakeAsyncLayer()
+    event_result = None
+
+    def on_done(event):
+        nonlocal event_result
+        event_result = event.value
+
+    layer_slicer.events.ready.connect(on_done)
+
+    future = layer_slicer.slice_layers_async(layers=[layer], dims=Dims())
+    actual_result = future.result()
+
+    assert actual_result is event_result
+
+
+def test_slice_layers_async_with_one_sync_layer(layer_slicer):
+    layer = FakeSyncLayer()
+    assert layer.slice_count == 0
+
+    future = layer_slicer.slice_layers_async(layers=[layer], dims=Dims())
+
+    assert layer.slice_count == 1
+    assert not future.result()
```

Comment:
Might be more unclear to avoid confusion with `None`?

```suggestion
    assert future.result() == {}
```